src/content/cookbook/architecture/result.md

```diff
@@ -0,0 +1,460 @@
+```dart
+Future<Result<UserProfile>> getUserProfile() async {
+  return await _apiClientService.getUserProfile();
+}
+```
+
+### Unwrapping the Result object
+
+Now the ViewModel doesn't receive the `UserProfile` directly, 
+but instead it receives a `Result` containing a `UserProfile`.
+
+This enforces the developer implementing the ViewModel 
+to unwrap the `Result` to obtain the `UserProfile`, 
+and avoids having uncaught exceptions.
+
+<?code-excerpt "lib/main.dart (UserProfileViewModel)"?>
+```dart
+class UserProfileViewModel extends ChangeNotifier {
+  // ···
+
+  UserProfile? userProfile;
+
+  Exception? error;
+
+  Future<void> load() async {
+    final result = await userProfileRepository.getUserProfile();
+    switch (result) {
+      case Ok<UserProfile>():
+        userProfile = result.value;
+      case Error<UserProfile>():
+        error = result.error;
+    }
+    notifyListeners();
+  }
+}
+```
+
+The `Result` class is implemented using a `sealed` class, 
+meaning it can only either be an `Ok` or an `Error`, 
+so this operation can be done using a switch case.
+
+In the `Ok<UserProfile>` case, 
+obtain the value using the `value` property. 
+
+In the `Error<UserProfile>` case,
+obtain the error object using the `error` property.
+
+## Improving control flow
+
+Wrapping code in a `try-catch` block
+is a quick way to ensure that  the calling method captures any thrown exception.
+Catching stops the exception from propagating to other methods, and it disrupts the normal code flow.
+
+Consider the following code.
+
+<?code-excerpt "lib/no_result.dart (UserProfileRepository2)" replace="/2//g"?>
+```dart
+class UserProfileRepository {
+  // ···
+
+  Future<UserProfile> getUserProfile() async {
+    try {
+      return await _apiClientService.getUserProfile();
+    } catch (e) {
+      try {
+        return await _databaseService.createTemporaryUser();
+      } catch (e) {
+        throw Exception('Failed to get user profile');
+      }
+    }
+  }
+}
+```
+
+In this method, the `UserProfileRepository` 
+attempts to obtain the `UserProfile` 
+using the `ApiClientService`.
+If it fails, it tries to create a temporary user in a `DatabaseService`.
+
+Because either Service methods can fail, 
+the code must catch the exceptions in both cases.
+
+This can be improved using the `Result` pattern:
+
+
+<?code-excerpt "lib/main.dart (getUserProfile)"?>
+```dart
+Future<Result<UserProfile>> getUserProfile() async {
+  final apiResult = await _apiClientService.getUserProfile();
+  if (apiResult is Ok) {
+    return apiResult;
+  }
+
+  final databaseResult = await _databaseService.createTemporaryUser();
+  if (databaseResult is Ok) {
+    return databaseResult;
+  }
+
+  return Result.error(Exception('Failed to get user profile'));
+}
+```
+
+In this case, if the `Result` object is an `Ok`, 
+then the function returns that object, 
+otherwise at the end of the function returns a `Result.Error` if none worked.
```

Comment:
```suggestion
In this code, if the `Result` object is an `Ok` instance, 
then the function returns that object; 
otherwise, it returns `Result.Error`.
```